rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -448,6 +464,217 @@ extern "C" rmw_ret_t rmw_set_log_severity(rmw_log_severity_t severity)
+
+  user_callback_data_t * data = &(srv->user_callback_data);
+
+  std::lock_guard<std::mutex> guard(data->mutex);
+
+  // Set the user callback data
+  data->callback = callback;
+  data->user_data = user_data;
+
+  if (callback) {
+    // Push events happened before having assigned a callback
+    for (size_t i = 0; i < data->unread_count; i++) {
+      callback(user_data);
+    }
+    data->unread_count = 0;
+  }
+
+  return RMW_RET_OK;
+}
+
+extern "C" rmw_ret_t rmw_client_set_listener_callback(
+  rmw_client_t * rmw_client,
+  rmw_listener_callback_t callback,
+  const void * user_data)
+{
+  auto cli = static_cast<CddsClient *>(rmw_client->data);
+
+  user_callback_data_t * data = &(cli->user_callback_data);
+
+  std::lock_guard<std::mutex> guard(data->mutex);
+
+  // Set the user callback data
+  data->callback = callback;
+  data->user_data = user_data;
+
+  if (callback) {
+    // Push events happened before having assigned a callback
+    for (size_t i = 0; i < data->unread_count; i++) {
+      callback(user_data);
+    }
+    data->unread_count = 0;
+  }
+
+  return RMW_RET_OK;
+}
+
+extern "C" rmw_ret_t rmw_guard_condition_set_listener_callback(
+  rmw_guard_condition_t * rmw_guard_condition,
+  rmw_listener_callback_t callback,
+  const void * user_data,
+  bool use_previous_events)
+{
+  auto gc = static_cast<CddsGuardCondition *>(rmw_guard_condition->data);
+
+  user_callback_data_t * data = &(gc->user_callback_data);
+
+  std::lock_guard<std::mutex> lock(data->mutex);
+
+  // Set the user callback data
+  data->callback = callback;
+  data->user_data = user_data;
+
+  if (callback && use_previous_events) {
+    // Push events happened before having assigned a callback
+    for (size_t i = 0; i < data->unread_count; i++) {
+      callback(user_data);
+    }
+    data->unread_count = 0;
+  }
+
+  return RMW_RET_OK;
+}
+
+template<typename T>
+static void event_set_listener_callback(
+  T event,
+  rmw_listener_callback_t callback,
+  const void * user_data,
+  bool use_previous_events)
+{
+  user_callback_data_t * data = &(event->user_callback_data);
+
+  std::lock_guard<std::mutex> guard(data->mutex);
+
+  // Set the user callback data
+  data->callback = callback;
+  data->event_data = user_data;
+
+  if (callback && use_previous_events) {
+    // Push events happened before having assigned a callback
+    for (size_t i = 0; i < data->events_unread_count; i++) {
+      callback(user_data);
+    }
+    data->events_unread_count = 0;
+  }
+}
+
+extern "C" rmw_ret_t rmw_event_set_listener_callback(
+  rmw_event_t * rmw_event,
+  rmw_listener_callback_t callback,
+  const void * user_data,
+  bool use_previous_events)
+{
+  switch (rmw_event->event_data_type) {
+    case RMW_SUBSCRIBER_EVENT:
+      {
+        auto sub_event = static_cast<CddsSubscription *>(rmw_event->data);
+        event_set_listener_callback(
+          sub_event, callback, user_data, use_previous_events);
```

Comment:
aren't you overriding the callback for other events? (and also for data)

Reply:
The `user_callback_data->callback`  is the same for all the listener callbacks (new data available or events), since they all do the same thing: push an event into the queue.
So yes, I'm overriding the callback but with the same function pointer it had before. This is because we don't know which api will be called, `rmw_subscription_set_listener_callback` or `event_set_listener_callback`, or maybe both. We have to set the callback in each of them.

Reply:
so that works for your specific use case, but it doesn't sounds like a good idea in general.
I would rather store one callback for each.

You're also overriding `event_data`, which sounds like the cause of this problem:

> When a reader is created we get a liveliness_changed event, which correctly calls it callback pushing an event on the queue of EventsExecutor. Then the waitable does the take_data() and calls rmw_take_event() API. The problem is when we get here the event appears as RMW_EVENT_REQUESTED_QOS_INCOMPATIBLE instead of RMW_EVENT_LIVELINESS_CHANGED. Due this reason, we have commented out the piece of code which assigns the liveliness_changed listener callback, until we identify what is the problem here.

Reply:
> it doesn't sounds like a good idea in general. I would rather store one callback for each.

Right, done in 520df09